File: NextApp/Android/core/data/src/test/kotlin/jp/komehyappyo/member/next/core/data/FirebaseRestCommunityRepositoryVideoQuestionTests.kt
package jp.komehyappyo.member.next.core.data

import jp.komehyappyo.member.next.core.model.DistributedVideo
import jp.komehyappyo.member.next.core.model.CommunityAdminAccess
import kotlinx.coroutines.runBlocking
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class FirebaseRestCommunityRepositoryVideoQuestionTests {
    @Test
    fun saveVideoQuestionUsesClientRequestIdAsIdempotentDocumentId() = runBlocking {
        installMockFirestoreStreamHandler()
        FakeFirestoreRequestHandler.reset()
        val path = "/v1/projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions"
        val request = PathMethodKey(
            path = path,
            query = "documentId=request-123",
            method = "POST",
        )
        val video = DistributedVideo(
            id = "video-1",
            communityId = "org-1",
            videoTitle = "動画",
            description = "",
            embedHtml = "",
            videoUrl = "",
            vimeoUrl = "",
            providerVideoId = "",
            videoType = "distributed_vimeo",
            thumbnailUrl = "",
            isPremium = false,
            createdAt = null,
            updatedAt = null,
            isPublished = true,
            isMembersOnly = false,
            sortOrder = 0,
        )

        listOf(200, 409).forEach { statusCode ->
            FakeFirestoreRequestHandler.responses = mapOf(
                request to FakeResponse(
                    statusCode = statusCode,
                    body = JSONObject().put("writeTime", JSONObject()).toString(),
                ),
            )
            repository().saveVideoQuestion(
                communityId = "org-1",
                memberUid = "member",
                video = video,
                memoText = "メモ",
                questionText = "質問",
                playbackSeconds = 10.0,
                clientRequestId = "request-123",
                idToken = "id-token",
            ).getOrThrow()
        }

        assertEquals(listOf(path, path), FakeFirestoreRequestHandler.requests.map { it.path })
        assertEquals(listOf("POST", "POST"), FakeFirestoreRequestHandler.requests.map { it.method })
        assertEquals(
            listOf("documentId=request-123", "documentId=request-123"),
            FakeFirestoreRequestHandler.requests.map { it.query },
        )
        val fields = JSONObject(FakeFirestoreRequestHandler.requests.first().body).getJSONObject("fields")
        assertEquals("request-123", fields.getJSONObject("clientRequestId").getString("stringValue"))
        assertEquals("synced", fields.getJSONObject("syncStatus").getString("stringValue"))
    }

    @Test
    fun adminVideoQuestionsLoadsAllQuestionsSortedByCreatedAt() = runBlocking {
        installMockFirestoreStreamHandler()
        FakeFirestoreRequestHandler.reset()

        FakeFirestoreRequestHandler.responses = mapOf(
            PathMethodKey(
                path = "/v1/projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions",
                query = "pageSize=1000",
                method = "GET",
            ) to FakeResponse(
                statusCode = 200,
                body = JSONObject().apply {
                    put(
                        "documents",
                        JSONArray().apply {
                            put(
                                JSONObject().apply {
                                    put("name", "projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions/q-new")
                                    put(
                                        "fields",
                                        JSONObject().apply {
                                            put("questionText", JSONObject().put("stringValue", "新しい質問"))
                                            put("videoTitle", JSONObject().put("stringValue", "動画A"))
                                            put("memoText", JSONObject().put("stringValue", "メモ"))
                                            put("answerText", JSONObject().put("stringValue", ""))
                                            put("memberUid", JSONObject().put("stringValue", "member-a"))
                                            put("videoId", JSONObject().put("stringValue", "video-a"))
                                            put("playbackSeconds", JSONObject().put("doubleValue", 10.0))
                                            put("createdAt", JSONObject().put("timestampValue", "2026-08-13T12:00:00Z"))
                                        },
                                    )
                                },
                            )
                            put(
                                JSONObject().apply {
                                    put("name", "projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions/q-old")
                                    put(
                                        "fields",
                                        JSONObject().apply {
                                            put("questionText", JSONObject().put("stringValue", "古い質問"))
                                            put("videoTitle", JSONObject().put("stringValue", "動画B"))
                                            put("answerText", JSONObject().put("stringValue", "既に回答"))
                                            put("memberUid", JSONObject().put("stringValue", "member-b"))
                                            put("videoId", JSONObject().put("stringValue", "video-b"))
                                            put("playbackSeconds", JSONObject().put("doubleValue", 20.0))
                                            put("createdAt", JSONObject().put("timestampValue", "2026-08-12T09:00:00Z"))
                                            put("answeredAt", JSONObject().put("timestampValue", "2026-08-12T10:00:00Z"))
                                        },
                                    )
                                },
                            )
                        },
                    )
                }.toString(),
            ),
        )

        val result = repository().adminVideoQuestions("org-1", "id-token").getOrThrow()

        assertEquals(2, result.size)
        assertEquals("q-new", result.first().id)
        assertEquals("q-old", result.last().id)
        assertEquals("動画A", result.first().videoTitle)
        assertEquals("member-a", result.first().memberUid)
        assertEquals("メモ", result.first().memoText)
        assertEquals("", result.first().answerText)
        assertEquals(null, result.first().answeredAt)
        assertEquals("2026-08-12T10:00:00Z", result.last().answeredAt)
    }

    @Test
    fun answerVideoQuestionSendsTrimmedPatchPayload() = runBlocking {
        installMockFirestoreStreamHandler()
        FakeFirestoreRequestHandler.reset()

        FakeFirestoreRequestHandler.responses = mapOf(
            PathMethodKey(
                path = "/v1/projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions/q-new",
                method = "PATCH",
            ) to FakeResponse(
                statusCode = 200,
                body = JSONObject().put("writeTime", JSONObject()).toString(),
            ),
        )

        val response = repository().answerVideoQuestion(
            "org-1",
            "q-new",
            "  回答します  ",
            "id-token",
        )

        assertTrue(response.isSuccess)
        assertEquals(1, FakeFirestoreRequestHandler.requests.size)
        val request = FakeFirestoreRequestHandler.requests.first()
        assertEquals("PATCH", request.method)
        assertEquals(
            "/v1/projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions/q-new",
            request.path,
        )
        val payload = JSONObject(request.body)
        val fields = payload.getJSONObject("fields")
        assertEquals("回答します", fields.getJSONObject("answerText").getString("stringValue"))
        assertTrue(fields.getJSONObject("updatedAt").has("timestampValue"))
        assertTrue(fields.getJSONObject("answeredAt").has("timestampValue"))
        assertEquals(
            fields.getJSONObject("updatedAt").getString("timestampValue"),
            fields.getJSONObject("answeredAt").getString("timestampValue"),
        )
    }

    @Test
    fun saveAdministratorPersistsSelectedPermissionSetAndAuditLog() = runBlocking {
        installMockFirestoreStreamHandler()
        FakeFirestoreRequestHandler.reset()
        FakeFirestoreRequestHandler.responses = mapOf(
            PathMethodKey(
                path = "/v1/projects/test-project/databases/(default)/documents:commit",
                method = "POST",
            ) to FakeResponse(
                statusCode = 200,
                body = JSONObject().put("commitTime", "2026-08-15T00:00:00Z").toString(),
            ),
        )

        repository().saveAdministrator(
            communityId = "org-1",
            adminUserId = " manager-1 ",
            role = "manager",
            permissions = setOf(
                CommunityAdminAccess.USAGE_ANALYTICS_READ_PERMISSION,
                CommunityAdminAccess.ANNOUNCEMENT_PUBLISH_PERMISSION,
            ),
            isActive = true,
            actorUserId = "owner-1",
            idToken = "id-token",
        ).getOrThrow()

        val request = FakeFirestoreRequestHandler.requests.single()
        assertEquals("POST", request.method)
        val writes = JSONObject(request.body).getJSONArray("writes")
        val administratorFields = writes.getJSONObject(0)
            .getJSONObject("update")
            .getJSONObject("fields")
        assertEquals(
            "projects/test-project/databases/(default)/documents/organizations/org-1/admins/manager-1",
            writes.getJSONObject(0).getJSONObject("update").getString("name"),
        )
        val permissions = administratorFields.getJSONObject("permissions")
            .getJSONObject("arrayValue")
            .getJSONArray("values")
        assertEquals(2, permissions.length())
        assertEquals("announcementPublish", permissions.getJSONObject(0).getString("stringValue"))
        assertEquals("usageAnalyticsRead", permissions.getJSONObject(1).getString("stringValue"))
        val auditFields = writes.getJSONObject(1).getJSONObject("update").getJSONObject("fields")
        assertEquals("administrator.added", auditFields.getJSONObject("action").getString("stringValue"))
        assertEquals("manager-1", auditFields.getJSONObject("targetUserId").getString("stringValue"))
    }

    private fun repository(): FirebaseRestCommunityRepository {
        return FirebaseRestCommunityRepository(projectId = "test-project")
    }
}

File: NextApp/iOS/OrgPortalNext/Tests/DataLayerTests/FirebaseRESTCommunityRepositoryVideoQuestionTests.swift
import Foundation
import Model
import XCTest
@testable import DataLayer

final class FirebaseRESTCommunityRepositoryVideoQuestionTests: XCTestCase {
    func testSaveVideoQuestionUsesClientRequestIdAsIdempotentDocumentId() async throws {
        MockFirestoreProtocol.reset()
        let session = URLSession(configuration: mockSessionConfiguration())
        let repository = FirebaseRESTCommunityRepository(projectId: "test-project", session: session)
        let path = "/v1/projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions"
        let request = requestKey(
            path: path,
            method: "POST",
            query: "documentId=request-123"
        )
        let video = DistributedVideo(
            id: "video-1",
            communityId: "org-1",
            videoTitle: "動画",
            description: "",
            embedHtml: "",
            videoUrl: "",
            vimeoUrl: "",
            providerVideoId: "",
            videoType: "distributed_vimeo",
            thumbnailUrl: "",
            isPremium: false,
            createdAt: nil,
            updatedAt: nil,
            isPublished: true,
            isMembersOnly: false,
            sortOrder: 0
        )

        for statusCode in [200, 409] {
            MockFirestoreProtocol.responses[request] = FakeResponse(
                statusCode: statusCode,
                body: ["writeTime": [:]]
            )
            try await repository.saveVideoQuestion(
                communityId: "org-1",
                memberUid: "member",
                video: video,
                memoText: "メモ",
                questionText: "質問",
                playbackSeconds: 10,
                clientRequestId: "request-123",
                idToken: "id-token"
            )
        }

        XCTAssertEqual(MockFirestoreProtocol.requests.map(\.path), [path, path])
        XCTAssertEqual(MockFirestoreProtocol.requests.map(\.method), ["POST", "POST"])
        XCTAssertEqual(
            MockFirestoreProtocol.requests.map(\.query),
            ["documentId=request-123", "documentId=request-123"]
        )
        let payload = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: MockFirestoreProtocol.requests[0].body) as? [String: Any]
        )
        let fields = try XCTUnwrap(payload["fields"] as? [String: Any])
        XCTAssertEqual(
            (fields["clientRequestId"] as? [String: Any])?["stringValue"] as? String,
            "request-123"
        )
        XCTAssertEqual(
            (fields["syncStatus"] as? [String: Any])?["stringValue"] as? String,
            "synced"
        )
    }

    func testAdminVideoQuestionsLoadsAllQuestionsSortedByCreatedAt() async throws {
        MockFirestoreProtocol.reset()
        let session = URLSession(
            configuration: mockSessionConfiguration()
        )
        let repository = FirebaseRESTCommunityRepository(
            projectId: "test-project",
            session: session,
        )

        MockFirestoreProtocol.responses[requestKey(
            path: "/v1/projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions",
            method: "GET",
            query: "pageSize=1000",
        )] = FakeResponse(
            statusCode: 200,
            body: [
                "documents": [
                    [
                        "name": "projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions/q-new",
                        "fields": [
                            "questionText": ["stringValue": "新しい質問"],
                            "videoTitle": ["stringValue": "動画A"],
                            "memoText": ["stringValue": "メモ"],
                            "answerText": ["stringValue": ""],
                            "memberUid": ["stringValue": "member-a"],
                            "videoId": ["stringValue": "video-a"],
                            "playbackSeconds": ["doubleValue": 10.0],
                            "createdAt": ["timestampValue": "2026-08-13T12:00:00Z"],
                        ],
                    ],
                    [
                        "name": "projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions/q-old",
                        "fields": [
                            "questionText": ["stringValue": "古い質問"],
                            "videoTitle": ["stringValue": "動画B"],
                            "answerText": ["stringValue": "既に回答"],
                            "memberUid": ["stringValue": "member-b"],
                            "videoId": ["stringValue": "video-b"],
                            "playbackSeconds": ["doubleValue": 20.0],
                            "createdAt": ["timestampValue": "2026-08-12T09:00:00Z"],
                            "answeredAt": ["timestampValue": "2026-08-12T10:00:00Z"],
                        ],
                    ],
                ],
            ],
        )

        let result = try await repository.adminVideoQuestions(
            communityId: "org-1",
            idToken: "id-token",
        )

        XCTAssertEqual(result.count, 2)
        XCTAssertEqual(result[0].id, "q-new")
        XCTAssertEqual(result[1].id, "q-old")
        XCTAssertEqual(result[0].videoTitle, "動画A")
        XCTAssertEqual(result[0].memberUid, "member-a")
        XCTAssertEqual(result[0].memoText, "メモ")
        XCTAssertEqual(result[0].answerText, "")
        XCTAssertNil(result[0].answeredAt)
        XCTAssertEqual(
            result[1].answeredAt,
            ISO8601DateFormatter().date(from: "2026-08-12T10:00:00Z")
        )
    }

    func testAnswerVideoQuestionSendsTrimmedPayload() async throws {
        MockFirestoreProtocol.reset()
        let session = URLSession(
            configuration: mockSessionConfiguration()
        )
        let repository = FirebaseRESTCommunityRepository(
            projectId: "test-project",
            session: session,
        )

        MockFirestoreProtocol.responses[requestKey(
            path: "/v1/projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions/q-new",
            method: "PATCH",
        )] = FakeResponse(
            statusCode: 200,
            body: ["writeTime": [:]],
        )

        _ = try await repository.answerVideoQuestion(
            communityId: "org-1",
            questionId: "q-new",
            answerText: "  回答します  ",
            idToken: "id-token",
        )

        XCTAssertEqual(MockFirestoreProtocol.requests.count, 1)
        let request = try XCTUnwrap(MockFirestoreProtocol.requests.first)
        XCTAssertEqual(request.method, "PATCH")
        XCTAssertEqual(
            request.path,
            "/v1/projects/test-project/databases/(default)/documents/organizations/org-1/videoQuestions/q-new"
        )

        let payload = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: request.body) as? [String: Any]
        )
        let fields = try XCTUnwrap(payload["fields"] as? [String: Any])
        let answerText = try XCTUnwrap((fields["answerText"] as? [String: Any])?["stringValue"] as? String)
        let updatedAt = (fields["updatedAt"] as? [String: Any])?["timestampValue"] as? String
        let answeredAt = (fields["answeredAt"] as? [String: Any])?["timestampValue"] as? String

        XCTAssertEqual(answerText, "回答します")
        XCTAssertNotNil(updatedAt)
        XCTAssertNotNil(answeredAt)
        XCTAssertEqual(answeredAt, updatedAt)
    }

    private func mockSessionConfiguration() -> URLSessionConfiguration {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.protocolClasses = [MockFirestoreProtocol.self]
        return configuration
    }
}

private struct RequestKey: Hashable {
    let path: String
    let method: String
    let query: String?

    init(path: String, method: String, query: String? = nil) {
        self.path = path
        self.method = method
        self.query = query
    }
}

private func requestKey(
    path: String,
    method: String,
    query: String? = nil,
) -> RequestKey {
    RequestKey(path: path, method: method, query: query)
}

private struct FakeResponse {
    let statusCode: Int
    let body: [String: Any]
}

private struct RecordedRequest {
    let method: String
    let path: String
    let query: String?
    let body: Data
}

private final class MockFirestoreProtocol: URLProtocol {
    nonisolated(unsafe) static var responses: [RequestKey: FakeResponse] = [:]
    nonisolated(unsafe) static var requests: [RecordedRequest] = []

    static func reset() {
        responses = [:]
        requests = []
    }

    override class func canInit(with request: URLRequest) -> Bool {
        true
    }

    override class func canonicalRequest(for request: URLRequest) -> URLRequest {
        request
    }

    override func startLoading() {
        let requestMethod = request.httpMethod ?? "GET"
        let requestPath = request.url?.path ?? ""
        let requestQuery = request.url?.query
        let key = requestKey(
            path: requestPath,
            method: requestMethod,
            query: requestQuery,
        )

        let response = Self.responses[key]
            ?? Self.responses[requestKey(path: requestPath, method: requestMethod)]
            ?? FakeResponse(statusCode: 500, body: ["error": "not found"])

        // URLSession.data(for:) converts httpBody into httpBodyStream before handing the
        // request to a custom URLProtocol, so fall back to draining the stream.
        let requestBody = request.httpBody ?? Self.readBody(from: request.httpBodyStream)
        Self.requests.append(
            RecordedRequest(
                method: requestMethod,
                path: requestPath,
                query: requestQuery,
                body: requestBody,
            )
        )

        guard let url = request.url else {
            let error = NSError(domain: "MockFirestoreProtocol", code: -1)
            client?.urlProtocol(self, didFailWithError: error)
            return
        }

        let responseData = try! JSONSerialization.data(withJSONObject: response.body)
        let httpResponse = HTTPURLResponse(
            url: url,
            statusCode: response.statusCode,
            httpVersion: "HTTP/1.1",
            headerFields: ["Content-Type": "application/json"],
        )!

        client?.urlProtocol(self, didReceive: httpResponse, cacheStoragePolicy: .notAllowed)
        client?.urlProtocol(self, didLoad: responseData)
        client?.urlProtocolDidFinishLoading(self)
    }

    private static func readBody(from stream: InputStream?) -> Data {
        guard let stream else { return Data() }
        stream.open()
        defer { stream.close() }
        var data = Data()
        let bufferSize = 4096
        var buffer = [UInt8](repeating: 0, count: bufferSize)
        while stream.hasBytesAvailable {
            let bytesRead = stream.read(&buffer, maxLength: bufferSize)
            if bytesRead <= 0 { break }
            data.append(buffer, count: bytesRead)
        }
        return data
    }

    override func stopLoading() {
        // No-op.
    }
}
